### PR TITLE
Use solana run-time for rbpf-tests and make a transaction with input data

### LINKED ISSUE #140: [Bug] no way to specify accounts used in tests
# 🐛 Bug
Solana on-chain programs can access only accounts explicitly specified as part of the setup of the transaction that invokes the on-chain program. Currently there’s no way to pass this information to the test harness system that runs executable on-chain programs.

## Expected Behavior

Extend test harness to accept additional
input file that describes account and instruction data to be passed to on-chain test programs.

### PR DESCRIPTION
**Pull request by dmakarov**
Resolves #140

### CODE REVIEW

File: language/tools/move-mv-llvm-compiler/tests/rbpf-tests.rs [hunk 1]
```diff
@@ -231,44 +250,178 @@ fn link_object_files(
...
-    use rbpf::{
-        ebpf, elf::Executable, error::EbpfError, memory_region::MemoryRegion,
-        verifier::RequisiteVerifier, vm::*,
+fn load_program<'a>(
+    filename: &Path,
+    program_id: Pubkey,
+    invoke_context: &InvokeContext<'a>,
+) -> Executable<RequisiteVerifier, InvokeContext<'a>> {
+    let contents = &std::fs::read(filename).unwrap_or_else(|e| {
+        eprintln!("Can't read the executbale {:?}, error: {}", filename, e);
```
Comment (nvjle) [inline language/tools/move-mv-llvm-compiler/tests/rbpf-tests.rs]: Micro nit "executable".

**Comment by nvjle** [comment]
For the patch itself, LGTM. 
Very nice work, we are very close now to be able to connect the input Solana account info to a Move function.

However, the lint-related checks fail. But these don't seem related to the patch per-se, other than additional crates are being pulled in. The custom `cargo x lint` fails due to a strange error which I wasn't able to easily diagnose (Move-blessed Rust version too old perhaps?). Also, `cargo xclippy --workspace --all-targets` fails with minor `unused imports` for a file in `move-prover`. Why that didn't fail before is not clear since that code wasn't touched. I don't understand how the custom Move x-tools work. But something is wonky with them.

**Comment by dmakarov** [comment]
I had to bump the version of one of the dependencies listed in x.toml because of a versions conflict. I’ll see how to fix the lint failures.

**Review by nvjle — APPROVED** [review]

**Comment by nvjle** [comment]
> I had to bump the version of one of the dependencies listed in x.toml because of a versions conflict. I’ll see how to fix the lint failures.

FWIW, I poked around and I think the problem might be crate `target-spec` 0.9 is too old to parse `target_abi`. But some package dependency is requiring that option on a cfg line. I looked at `cargo tree` and can't quite figure out the dependence chain leading to that (even with --depth option). That's about the extent of my cargo kung fu.

**Comment by nvjle** [comment]
> > I had to bump the version of one of the dependencies listed in x.toml because of a versions conflict. I’ll see how to fix the lint failures.
> 
> FWIW, I poked around and I think the problem might be crate `target-spec` 0.9 is too old to parse `target_abi`. But some package dependency is requiring that option on a cfg line. I looked at `cargo tree` and can't quite figure out the dependence chain leading to that (even with --depth option). That's about the extent of my cargo kung fu.

I think I've mostly figured it out. 
- One of the new Solana dependencies transitively pulls in `windows-targets 0.48.0`.
- The above depends on rust/cargo being able to parse a cfg line with `target_abi`.
- Those cfg lines are parsed by crates `cfg-expr` + `target-spec`, and these need to be relatively recent to parse `target_abi`.
- The weird custom diem tools in `devtools/x*` depend on crate `guppy 0.12.3`, and that crate depends on `target-spec 0.9.0`-- too old to parse `target_abi`.
- Fixing the previous requires updating to `guppy 0.15.2`, which depends on a newer `target-spec 1.3.0` which successfully parses `target_abi`.
- Updating the last line requires updating the Move-blessed rust toolchain from 1.65 to at least 1.66 (`move/rust-toolchain`).
- But one of the guppy APIs the custom diem tools uses has changed since 0.12.3, so that code needs updating. From a cursory look, I think it is just one method that has either changed names or moved from one class to another.
- Also, the weird custom diem tools depend on a specific revision of custom crates `nextest-runner` and `nextest-config`. The first also needs its Cargo.toml updated for the newer guppy. We'd either have to fork our own copy of that specific repository and update the toml, or submit a PR to the crate owner.

This cargo + weird custom diem tools creates a surprisingly convoluted and delicate house of cards. And I couldn't get any of the tools to tell me the above, which one would imagine it should easily do. What a pain.

**Comment by dmakarov** [comment]
> > > I had to bump the version of one of the dependencies listed in x.toml because of a versions conflict. I’ll see how to fix the lint failures.
> > 
> > 
> > FWIW, I poked around and I think the problem might be crate `target-spec` 0.9 is too old to parse `target_abi`. But some package dependency is requiring that option on a cfg line. I looked at `cargo tree` and can't quite figure out the dependence chain leading to that (even with --depth option). That's about the extent of my cargo kung fu.
> 
> I think I've mostly figured it out.
> 
>     * One of the new Solana dependencies transitively pulls in `windows-targets 0.48.0`.
> 
>     * The above depends on rust/cargo being able to parse a cfg line with `target_abi`.
> 
>     * Those cfg lines are parsed by crates `cfg-expr` + `target-spec`, and these need to be relatively recent to parse `target_abi`.
> 
>     * The weird custom diem tools in `devtools/x*` depend on crate `guppy 0.12.3`, and that crate depends on `target-spec 0.9.0`-- too old to parse `target_abi`.
> 
>     * Fixing the previous requires updating to `guppy 0.15.2`, which depends on a newer `target-spec 1.3.0` which successfully parses `target_abi`.
> 
>     * Updating the last line requires updating the Move-blessed rust toolchain from 1.65 to at least 1.66 (`move/rust-toolchain`).
> 
>     * But one of the guppy APIs the custom diem tools uses has changed since 0.12.3, so that code needs updating. From a cursory look, I think it is just one method that has either changed names or moved from one class to another.
> 
>     * Also, the weird custom diem tools depend on a specific revision of custom crates `nextest-runner` and `nextest-config`. The first also needs its Cargo.toml updated for the newer guppy. We'd either have to fork our own copy of that specific repository and update the toml, or submit a PR to the crate owner.
> 
> 
> This cargo + weird custom diem tools creates a surprisingly convoluted and delicate house of cards. And I couldn't get any of the tools to tell me the above, which one would imagine it should easily do. What a pain.

Thank you for the very thorough analysis.  I was trying to get around this issue, by disabling default features of the chrono crate. The windows-target crate is pulled indirectly by one of the optional dependencies of the updated chrono crate (required by solana-* dependencies). This optional dependency (iana-time-zone) shouldn't be pulled in for "std" feature (it's required only for "clock" feature of chrono, which I don't think is used). However, cargo stubbornly pulls in all optional dependencies regardless of the features I specify for the chrono crate. I'll follow your analysis, and see how to fix this, probably early next week.

**Comment by nvjle** [comment]
> > > > I had to bump the version of one of the dependencies listed in x.toml because of a versions conflict. I’ll see how to fix the lint failures.
> > > 
> > > 
> > > FWIW, I poked around and I think the problem might be crate `target-spec` 0.9 is too old to parse `target_abi`. But some package dependency is requiring that option on a cfg line. I looked at `cargo tree` and can't quite figure out the dependence chain leading to that (even with --depth option). That's about the extent of my cargo kung fu.
> > 
> > 
> > I think I've mostly figured it out.
> > ```
> > * One of the new Solana dependencies transitively pulls in `windows-targets 0.48.0`.
> > 
> > * The above depends on rust/cargo being able to parse a cfg line with `target_abi`.
> > 
> > * Those cfg lines are parsed by crates `cfg-expr` + `target-spec`, and these need to be relatively recent to parse `target_abi`.
> > 
> > * The weird custom diem tools in `devtools/x*` depend on crate `guppy 0.12.3`, and that crate depends on `target-spec 0.9.0`-- too old to parse `target_abi`.
> > 
> > * Fixing the previous requires updating to `guppy 0.15.2`, which depends on a newer `target-spec 1.3.0` which successfully parses `target_abi`.
> > 
> > * Updating the last line requires updating the Move-blessed rust toolchain from 1.65 to at least 1.66 (`move/rust-toolchain`).
> > 
> > * But one of the guppy APIs the custom diem tools uses has changed since 0.12.3, so that code needs updating. From a cursory look, I think it is just one method that has either changed names or moved from one class to another.
> > 
> > * Also, the weird custom diem tools depend on a specific revision of custom crates `nextest-runner` and `nextest-config`. The first also needs its Cargo.toml updated for the newer guppy. We'd either have to fork our own copy of that specific repository and update the toml, or submit a PR to the crate owner.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This cargo + weird custom diem tools creates a surprisingly convoluted and delicate house of cards. And I couldn't get any of the tools to tell me the above, which one would imagine it should easily do. What a pain.
> 
> Thank you for the very thorough analysis. I was trying to get around this issue, by disabling default features of the chrono crate. The windows-target crate is pulled indirectly by one of the optional dependencies of the updated chrono crate (required by solana-* dependencies). This optional dependency (iana-time-zone) shouldn't be pulled in for "std" feature (it's required only for "clock" feature of chrono, which I don't think is used). However, cargo stubbornly pulls in all optional dependencies regardless of the features I specify for the chrono crate. I'll follow your analysis, and see how to fix this, probably early next week.

It was a worthwhile exercise. Even if we managed to have some trivial work around now, I doubt we will sidestep some of these issues for very long (i.e., cfg-expr/target-spec seem to be used somewhat frequently and guppy too). Also, it gave me a chance to answer my own question I asked about a few weeks ago-- can we update `move/rust-toolchain` without the world collapsing, and the answer is yes. That done is isolation did work and all of our move-mv-llvm-compiler tests still pass.

**Pull request merged**